### PR TITLE
Adding dnn based super resolution module.

### PR DESCRIPTION
**Pull request by fannymonori**
**Merge with extra:** opencv/opencv_extra#662

Final and complete PR of GSoC 2019, project of Fanny Monori. Parts of the code was done with Xavier Weber, who worked on the same GSoC project. His PR is available here: opencv/opencv_contrib#2231

My own contribution was implementing the [ESPCN](https://github.com/fannymonori/TF-ESPCN) and [LapSRN](https://github.com/fannymonori/TF-LapSRN) code in TensorFlow, doing parts of the main functionality, do the benchmarking functionality, and corresponding tutorials, samples, and tests. However we shared most of the work with the main functionality  (importing the models and doing the super-resolution).

**This pull request changes:**

- Adds the dnn based super resolution modules main functionality
- Benchmarking functionality for evaluating the DL models and comparing them to existing interpolation methods
- Adds sample codes, tutorials, and unit tests.

### CODE REVIEW

**Comment by alalek** [comment]
Thank you for the contribution!

Please ask your mentor to take a look and approve PR changes.

File: modules/dnn_superres/include/opencv2/dnn_superres_quality.hpp [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+#include <opencv2/highgui.hpp>
+#include <opencv2/imgproc.hpp>
+#include <cstdarg>
+#include <sstream>
+#include <ctime>
+
+#ifdef _WIN32
+#include <Windows.h>
+#endif
```
Comment (alalek) [inline modules/dnn_superres/include/opencv2/dnn_superres_quality.hpp]: Do we really need platform specific include in public includes?

> #include <opencv2/highgui.hpp>
> ...

A lot of these headers are not needed to define this API.
Move these includes into corresponding .cpp files.

File: modules/dnn_superres/src/dnn_superres.cpp [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+#include "precomp.hpp"
+
+#include "opencv2/dnn_superres.hpp"
+
+namespace cv
+{
+    namespace dnn_superres
+    {
+
+        int DnnSuperResImpl::layer_loaded = 0;
```
Comment (alalek) [inline modules/dnn_superres/src/dnn_superres.cpp]: Please avoid unnecessary indentation in namespaces.

File: modules/dnn_superres/src/dnn_superres_quality.cpp [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+            #ifdef _WIN32
+                LARGE_INTEGER st, et, el, fq;
+            #else
+                double start, end;
+                timespec ts_beg, ts_end;
+            #endif
+
+            #ifdef _WIN32
+                QueryPerformanceFrequency(&fq);
+                QueryPerformanceCounter(&st);
```
Comment (alalek) [inline modules/dnn_superres/src/dnn_superres_quality.cpp]: Use `int64 getTickCount();` or `TickMeter` helper class instead:
https://github.com/opencv/opencv/blob/4.1.1/modules/core/include/opencv2/core/utility.hpp#L285-L332

File: modules/dnn_superres/src/dnn_superres_quality.cpp [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+        {
+            fontScale = fontscale;
+        }
+
+        void DnnSuperResQuality::setFontColor(cv::Scalar fontcolor)
+        {
+            fontColor = fontcolor;
+        }
+
+        double DnnSuperResQuality::psnr(Mat img, Mat orig)
```
Comment (alalek) [inline modules/dnn_superres/src/dnn_superres_quality.cpp]: Please use `PSNR()` function from core module instead

File: modules/dnn_superres/src/dnn_superres_quality.cpp [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+            double mse = rgb_sum / (double) (3.0 * img_float.total());
+
+            double max = 255 * 255;
+            double psnr = 10 * log10(max / (double) mse);
+
+            return psnr;
+        }
+
+        double DnnSuperResQuality::ssim(Mat img, Mat orig)
+        {
```
Comment (alalek) [inline modules/dnn_superres/src/dnn_superres_quality.cpp]: Did you take a look on SSIM from `quality` module (opencv_contrib)?

File: modules/dnn_superres/src/dnn_superres_quality.cpp [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+            int width = img.cols - (img.cols % scale);
+            int height = img.rows - (img.rows % scale);
+            Mat cropped = img(Rect(0, 0, width, height));
+            Mat imgDownscaled;
+
+            resize(cropped, imgDownscaled, Size(), 1.0/scale, 1.0/scale);
+
+            Mat imgUpscaled;
+
+			double elapsed = 0.0;
```
Comment (alalek) [inline modules/dnn_superres/src/dnn_superres_quality.cpp]: avoid "tabs". Indentation is 4 spaces

File: modules/dnn_superres/src/dnn_superres_quality.cpp [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+            Mat ssim = nominator / denominator;
+
+            Scalar mssim = mean(ssim);
+
+            return mssim[0];
+        }
+
+        void DnnSuperResQuality::benchmark(DnnSuperResImpl sr, Mat img,
+                                            std::vector<double>& psnrValues,
+                                            std::vector<double>& ssimValues,
```
Comment (alalek) [inline modules/dnn_superres/src/dnn_superres_quality.cpp]: I believe this should be moved into a separate sample (module itself is not good place for functions with a lot of customization parameters and/or drawing).

For reproducible benchmark results (in meaning of time) consider using of performance tests which collects multiple samples and drop outliers automatically.

File: modules/dnn_superres/test/test_dnn_superres.cpp [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+        void CV_DnnSuperResSingleOutputTest::run(int)
+        {
+            //x2
+            runOneModel("espcn", 2, "ESPCN_x2.pb");
+        }
+
+        TEST(CV_DnnSuperResSingleOutputTest, accuracy)
+        {
+            CV_DnnSuperResSingleOutputTest test;
+            test.safe_run();
```
Comment (alalek) [inline modules/dnn_superres/test/test_dnn_superres.cpp]: We don't need testing class here and below (it is from legacy testing system).
Just call `runOneModel(...)` directly (or move its content into test):
```
TEST(DnnSuperResSingleOutputTest, accuracy)
{
    cv::Ptr<DnnSuperResImpl> dnn_sr = makePtr<DnnSuperResImpl>();
    ... all your test code from runOneModel() ...
}
```

File: modules/dnn_superres/src/dnn_superres.cpp [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+        void DnnSuperResImpl::upsampleMultioutput(Mat img, std::vector<Mat> &imgs_new, std::vector<int> scale_factors, std::vector<String> node_names)
+        {
+            CV_Assert(scale_factors.size() == node_names.size());
+            CV_Assert(!scale_factors.empty());
+            CV_Assert(!node_names.empty());
+
+            if ( this->alg != "lapsrn" )
+            {
+                std::cout << "Only LapSRN support multiscale upsampling for now!" << std::endl;
+                return;
```
Comment (alalek) [inline modules/dnn_superres/src/dnn_superres.cpp]: Please use `CV_Error()` instead of stdout.

File: modules/dnn_superres/src/dnn_superres.cpp [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+
+        void DnnSuperResImpl::upsampleVideo(String inputPath, String outputPath)
+        {
+            VideoCapture inputVideo(inputPath);
+            int ex = static_cast<int>(inputVideo.get(CAP_PROP_FOURCC));
+            Size S = Size((int) inputVideo.get(CAP_PROP_FRAME_WIDTH) * this->sc,
+                            (int) inputVideo.get(CAP_PROP_FRAME_HEIGHT) * this->sc);
+
+            VideoWriter outputVideo;
+            outputVideo.open(outputPath, ex, inputVideo.get(CAP_PROP_FPS), S, true);
```
Comment (alalek) [inline modules/dnn_superres/src/dnn_superres.cpp]: This code of writer creation doesn't work well.

Do we really need this method in the module code? What is about cameras?
Does it make sense? Video compression algorithms can significantly amend results of algorithms.

File: modules/dnn_superres/include/opencv2/dnn_superres_quality.hpp (hunk 1 above)
Comment (fannymonori) [reply modules/dnn_superres/include/opencv2/dnn_superres_quality.hpp]: Corrected. I also don't need platform specific include anymore, and eliminated it.

File: modules/dnn_superres/src/dnn_superres.cpp (hunk 2 above)
Comment (fannymonori) [reply modules/dnn_superres/src/dnn_superres.cpp]: Done, I reformatted the code.

File: modules/dnn_superres/src/dnn_superres_quality.cpp (hunk 3 above)
Comment (fannymonori) [reply modules/dnn_superres/src/dnn_superres_quality.cpp]: Changed it to TickMeter.

File: modules/dnn_superres/src/dnn_superres_quality.cpp (hunk 4 above)
Comment (fannymonori) [reply modules/dnn_superres/src/dnn_superres_quality.cpp]: Replaced it with core module function.

File: modules/dnn_superres/src/dnn_superres_quality.cpp (hunk 5 above)
Comment (fannymonori) [reply modules/dnn_superres/src/dnn_superres_quality.cpp]: Replaced it with using quality module implementation.

File: modules/dnn_superres/src/dnn_superres_quality.cpp (hunk 6 above)
Comment (fannymonori) [reply modules/dnn_superres/src/dnn_superres_quality.cpp]: Done. Reformatted the code to match the guideline more.

File: modules/dnn_superres/src/dnn_superres_quality.cpp (hunk 7 above)
Comment (fannymonori) [reply modules/dnn_superres/src/dnn_superres_quality.cpp]: I removed it from the module itself, and moved it to just sample codes. I agree that it's better that way. And I added performance test to the module.

File: modules/dnn_superres/test/test_dnn_superres.cpp (hunk 8 above)
Comment (fannymonori) [reply modules/dnn_superres/test/test_dnn_superres.cpp]: I updated the test codes.

File: modules/dnn_superres/src/dnn_superres.cpp (hunk 9 above)
Comment (fannymonori) [reply modules/dnn_superres/src/dnn_superres.cpp]: I replaced them with CV_Error.

File: modules/dnn_superres/src/dnn_superres.cpp (hunk 10 above)
Comment (fannymonori) [reply modules/dnn_superres/src/dnn_superres.cpp]: Removed it from part of the module into just sample code, it makes more sense there.

**Pull request merged**